Fix timeframe bugs

gomorizsolt opened the pull request:
Issue #67.

gomorizsolt committed:
Refactor, respect timeframe when aggregating calendars

gomorizsolt commented on src/TeamContributionCalendar/TeamContributionCalendar.js:
```diff
@@ -140,10 +143,18 @@ export default class TeamContributionCalendar {
+              ? day
+              : undefined
+          )
+        );
```
It's a draft, tests fail, but let's discuss the changes first. My idea goes as follows:
- the structure of the calendars should be simplified somehow(see `sanitized()`, this basically removes the "noise" from the svg, e.g. attributes, month and weekday texts around the calendar etc)
- the calendars should be narrowed down according to the timeframe right before passing them to the `process*` functions
- `undefined` should be returned in the case of GitHub calendars to update the daily contributions at the correct position

gomorizsolt committed:
Re-add getCalendarDataByIndexes() call

gomorizsolt committed:
Respect timeframe and contributions

gomorizsolt committed:
Add clarification comments, simplify flow

gomorizsolt committed:
Remove comment

thisismydesign commented on src/utils/GitHubUtils/GitHubUtils.js:
```diff
@@ -61,56 +61,58 @@ export const setEmptyCalendarValues = calendar => {
+        );
+
+        const totalDailyContributions =
+          +actualCalendarDailyData.attributes["data-count"] +
```
The explicit cast to Number was easier to read imo.

gomorizsolt replied:
https://github.com/c-hive/team-contribution-calendar/pull/70/commits/dcd7948a18ff70fef7737a8f88f2fac8fe5feb47

thisismydesign commented on src/TeamContributionCalendar/TeamContributionCalendar.js:
```diff
@@ -140,10 +143,19 @@ export default class TeamContributionCalendar {
-          end: user.to
-        });
+        const timeframe = { start: user.from, end: user.to };
+        // This removes the "noise" from the calendar and sorts the days out falling out of the specified timeframe.
```
"sorts the days out falling out of the specified timeframe" does this mean "removes days falling out of the specified timeframe"?

thisismydesign replied:
I found this a bit weird :D Could you use the second form?

gomorizsolt replied:
Yea, sure.

https://github.com/c-hive/team-contribution-calendar/pull/70#discussion_r408411065

We may be able to get rid of these comments if there's a way to wrap the logic with a function.

gomorizsolt replied:
Obsolete.

thisismydesign commented on src/TeamContributionCalendar/TeamContributionCalendar.js:
```diff
@@ -159,22 +171,30 @@ export default class TeamContributionCalendar {
-        });
+        const timeframe = { start: user.from, end: user.to };
+        // This sorts the days out falling out of the specified timeframe and restores the array of arrays structure(i.e. [[date, contributions], ...]) to key-value pairs of an object.
+        const filteredCalendar = Object.entries(data.parsedCalendar)
```
How about creating a function for this that explains instead of the comment? Same for the above. If it's too complex to simply name and understand let's extract it and write specs for it.

gomorizsolt replied:
What do you mean by "write specs for it"?

thisismydesign replied:
specs == tests

gomorizsolt replied:
Obsolete.

thisismydesign commented on src/utils/GitHubUtils/GitHubUtils.js:
```diff
@@ -61,56 +61,58 @@ export const setEmptyCalendarValues = calendar => {
+  calendar.children[0].children
+    .filter(item => {
+      const isDay =
+        item.attributes.class !== "wday" && item.attributes.class !== "month";
```
Isn't there a specific class for days that you can check for? This way if they introduce a new class this would break.

Also why isn't `wday` a day? :)

gomorizsolt replied:
Let's first understand the structure then.

![Screenshot 2020-04-16 at 10 46 50](https://user-images.githubusercontent.com/30864202/79435163-9bf98400-7fcf-11ea-8566-9512b19b7d1d.png)

The "item" (from `filter(item => ...`) refers to either a "week" or a text around the calendar (text == months on the top and days on the left side). 

![Screenshot 2020-04-16 at 10 50 56](https://user-images.githubusercontent.com/30864202/79435593-2a6e0580-7fd0-11ea-9921-091e659a9e9b.png)

Each "week" has a `children` prop which represents the days of "that" week. For example, the 52nd element refers to the last week, i.e. the last column within the calendar; this week has 5 children element, that's why the last column consists of 5 squares (as opposed to the column to it which have 7 children consistently).

Starting from the 53rd item, they no longer refer to "weeks"; those items are the months and days around the calendar even though they have a `children` prop as well (but this prop always consists of a single "text" element). What makes them distinguishable from weeks is the "class" attribute they have (`month` for months and `wday` for the days on the left side).

As for the questions:

> Isn't there a specific class for days that you can check for? This way if they introduce a new class this would break.

We may use the `name` attribute because that's also different (it's `g` in the case of weeks whereas it's `element` in the case of the texts). But I'd say using the `attributes.class` to differentiate between "genuine" weeks and texts is the ideal approach.

> Also why isn't wday a day? :)

The clarification above should answer this question.

thisismydesign replied:
Thanks, it's clear now. I'd still propose to do a positive check instead of a negative one. So either check for the name or perhaps select the first 52 elements (if that's a constant pattern). Regarding whether this is needed or not, I think this is just a filter that might not be needed at all, for example, if you iterate on weeks based on indexes you don't need to care about indexes higher than 53. This is what I meant here: https://github.com/c-hive/team-contribution-calendar/pull/70#discussion_r408412767

thisismydesign replied:
If this is needed, I'd use `filter` that is more expressive in this case imo.

gomorizsolt replied:
> If this is needed, I'd use filter that is more expressive in this case imo.

We're already using `filter()` but it's combined with `map()` to convert the object of arrays to an array of arrays structure. Perhaps the conversion isn't necessary as removing the "noise" is a fairly nice improvement per se.

> So either check for the name or perhaps select the first 52 elements (if that's a constant pattern).

It appears to be a constant pattern. Then theoretically we can use `slice()` to cut off the elements that are supposed to be disregarded (i.e. the months and days).

gomorizsolt replied:
> This results in the exact same structure as currently except if the array's length varies.

On second thoughts, it's not the exact same structure but I'll look into it.

thisismydesign replied:
> We're already using filter() but it's combined with map() to convert the array of objects to an array of arrays structure.

Either way you could just add another fliter.

gomorizsolt replied:
https://github.com/c-hive/team-contribution-calendar/pull/70#issuecomment-615214224

thisismydesign commented on src/utils/GitHubUtils/GitHubUtils.js:
```diff
@@ -61,56 +61,58 @@ export const setEmptyCalendarValues = calendar => {
-  gitHubUserJsonCalendar,
-  timeframe
-) => {
+export const sanitize = calendar =>
```
> the structure of the calendars should be simplified somehow(see sanitized(), this basically removes the "noise" from the svg, e.g. attributes, month and weekday texts around the calendar etc)

Note that this introduces noise in the code instead. Is this necessary? I'd prefer simpler code if it can handle "noisy" data.

gomorizsolt replied:
> Note that this introduces noise in the code instead.

I don't think so. Why would we "ditch" the logic if it's way simpler to work with this structure?

gomorizsolt replied:
I'd propose discussing it.

thisismydesign replied:
> it's way simpler to work with this structure

if that's the case then sure it makes sense

gomorizsolt committed:
Use Number() to explicitly cast strings to numbers

gomorizsolt committed:
Revolutionize the flow, moving code around, project cleanup

gomorizsolt committed:
Resolve merge conflicts

gomorizsolt commented:
After discussing the threads above, I've decided to "revolutionize" the flow we've been stuck with since the beginnings. Highlights:
- the structure of the calendars is identical, consider the weights it takes off of our shoulds:
   - it's no longer complicated to iterate over the weeks/days (the nesting is inevitable though)
   - it's no longer required to have different `process*` functions
   - it's no longer required to have different "aggregation" functions
- adhering to immutability was always been tough because of having deeply nested calendars
   - https://github.com/immerjs/immer => installed as a dependency
   - it's no longer required to "worry about" mutating objects in an immutable manner

Let's discuss if needed. I'm open to suggestions. I'm pretty sure these changes make sense even if you're against it.

gomorizsolt commented on src/utils/GitHubUtils/GitHubUtils.js:
```diff
@@ -28,88 +28,45 @@ const getUserSvg = async (proxyServerUrl, gitHubUsername) => {
+        "data-count"
+      ] = 0;
+      draft.children[0].children[weekIndex].children[dayIndex].attributes.fill =
+        "#ebedf0";
```
Reason for this diff: I don't remember off the top of my head why, but we iterated over the calendar twice (now there's only a single, nested iteration).

gomorizsolt commented on src/utils/GitHubUtils/GitHubUtils.js:
```diff
@@ -28,88 +28,45 @@ const getUserSvg = async (proxyServerUrl, gitHubUsername) => {
-};
+export const dailyDataWithContributionsTransformation = calendar => {
+  // Extract the weeks from the calendar.
+  const weeks = calendar.children[0].children.slice(0, 53);
```
One thing that'd make sense here is to extract the "magical" `53` number to a variable.

thisismydesign replied:
Yes please, also shouldn't it be 52?

thisismydesign replied:
Also same as above:

> This comment can very easily be replaced by a function. Always think if comments are really necessary.

The logic here looks very similar to `CalendarUtils`, couldn't that be used here?

gomorizsolt replied:
> Yes please, also shouldn't it be 52?

No, it's 53 (52 weeks + the current week).

> The logic here looks very similar to CalendarUtils, couldn't that be used here?

Hmmm, they're identical in that the both of them extract the first 53 elements from the calendar and iterates over these elements(two iterations, one for the weeks and one for the days). The difference lies in what they're doing with the weeks/days in each iteration.

Extracting the iterations to a function might make sense:

```js
const iterateCalendar = (calendar, cb) => {
   const weeks = extractWeeks(calendar);

   weeks.forEach(week => {
      week.children.forEach(day => {
         cb(day);
      });
   });
};

iterateCalendar(calendar, day => { ... });
```


But (IMO) it's just rather a "wrapper" and a "shorthand" version which replaces the double invoked `forEach()` calls.

thisismydesign replied:
Fair enough

gomorizsolt committed:
Improve namings

gomorizsolt commented on src/utils/CalendarUtils/CalendarUtils.js:
```diff
@@ -37,30 +38,67 @@ export const elementExists = selector => {
+    }
+  );
+
+  return filtered.reduce(
```
We may rather [`Object.fromEntries()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/fromEntries) to transform the array of arrays into an object. My concerns are that [certain versions of browsers](https://caniuse.com/#search=fromEntries) don't yet support it(e.g. IE, what a surprise) and the tests failed because the runner couldn't recognize this function for some reason.

gomorizsolt replied:
Speaking of browser support: https://medium.com/hackernoon/introducing-immer-immutability-the-easy-way-9d73d8f71cb3 (see "No proxies?" chapter).

thisismydesign replied:
Looking at the "usage relative" chart it's ok t use imo. However, wouldn't our build process take care of things like these? Don't we transpile the code to work from most browsers?

> see "No proxies?" chapter
I don't know what proxies are in this context.

gomorizsolt replied:
> However, wouldn't our build process take care of things like these?

`Object.fromEntries()` is *not* transpiled after running either the `lib` or `distify` script.

> Don't we transpile the code to work from most browsers?

You're right, the target browsers are explicitly set:

https://github.com/c-hive/team-contribution-calendar/blob/f6ce3a54aab963a1b4584cdc574e39fee6fa599a/package.json#L86-L94

It should answer why it's not transpiled.

As for the error raised by tests: switching to a recent version of node(>=12.4.0) could resolve the issue (https://node.green/#ES2019).

gomorizsolt replied:
> switching to a recent version of node(>=12.4.0)

Would it make sense to specify the minimum node version?

gomorizsolt replied:
☝️ https://github.com/c-hive/team-contribution-calendar/pull/70/commits/4096ca7b250ee1f950a62b0a18633b48a66ee07e

thisismydesign replied:
:+1:

thisismydesign replied:
Would it make sense to set the `browserslist` to a broader one and include IE? If there's no noticeable downside we could do that.

gomorizsolt replied:
> Would it make sense to set the browserslist to a broader one and include IE?

Yea, it'd. But `Object.entries()` is still *not* transpiled for some reason.

https://github.com/babel/babel/issues/8362#issuecomment-407086767

It says we should use a "polyfill" to transpile methods.

https://babeljs.io/docs/en/babel-polyfill

> 🚨 As of Babel 7.4.0, this package has been deprecated in favor of directly including core-js/stable (to polyfill ECMAScript features) and regenerator-runtime/runtime (needed to use transpiled generator functions):

What this actually indicates is that we can directly import the "only-required-for-you polyfills" at the top of the entry point:

```js
import "core-js/stable";
import "regenerator-runtime/runtime";
```

One another, and way more preferred workaround to set up the polyfill, is:

https://github.com/zloirock/core-js#babelpreset-env

> @babel/preset-env has useBuiltIns option, which optimizes working with global version of core-js. With useBuiltIns option, you should also set corejs option to used version of core-js, like corejs: '3.6'.

So something like this:

```
// .babelrc

{
   "presets": [
      "@babel/preset-env,
      {
         // adds to the top of each file import of polyfills for features used in this file and not supported by target environments, so for
         "useBuiltIns": "entry"
      }
   ]
}
```

One another thing I've "observed" is that the `browserslist` entry isn't picked up by Babel in production whatsoever:

https://github.com/c-hive/team-contribution-calendar/blob/f6ce3a54aab963a1b4584cdc574e39fee6fa599a/package.json#L87

https://github.com/browserslist/browserslist#best-practices

> There is a defaults query, which gives a reasonable configuration for most users:

I'd fall back to their reasonable set of defaults for both development and production.

gomorizsolt replied:
> One another thing I've "observed" is that the browserslist entry isn't picked up by Babel in production whatsoever:

Ahh, now I see. So IE 11 is excluded from the list of browsers because it doesn't even support e.g. [async functions](https://caniuse.com/#search=async).

https://github.com/babel/babel/issues/9849#issuecomment-487040428

> To fix this, simply use regenerator-runtime, following the instructions in its README to ensure regeneratorRuntime is made available globally.

I don't think it's worth adding yet another dependency to make the package available for a relatively small subset of users (1.43%).

gomorizsolt replied:
https://github.com/c-hive/team-contribution-calendar/pull/70/commits/a120cad1391ec4f2c750fcd9b01e19a6f9ffc095

gomorizsolt replied:
> One another thing I've "observed" is that the browserslist entry isn't picked up by Babel in production whatsoever:

One more thing to add to this is that `browserslist` will fall back to the `defaults` if the target environment isn't specified:

> If none of them is declared, Browserslist will firstly look for production queries and then use defaults.

thisismydesign replied:
> I don't think it's worth adding yet another dependency to make the package available for a relatively small subset of users (1.43%).

These are all dev dependencies though, right? If there's a configuration that isn't too complicated that allows us to have a bigger reach, let's use it. If not, 1.4% is not a bug issue.

thisismydesign replied:
> not dead with a usage number like > 0.2%

This sounds reasonable to me.

thisismydesign replied:
I don't get the fallbacks in different envronments. "target environment isn't specified" => How do you specify it? Why isn't it specified?

gomorizsolt replied:
> These are all dev dependencies though, right? If there's a configuration that isn't too complicated that allows us to have a bigger reach, let's use it. If not, 1.4% is not a bug issue.

Yep, AFAICU it's merely a dev dependency and a one-line import at the top of the entry file.

> How do you specify it?

Target environment == "target browsers". These can be defined one-by-one by editing the `browserslist` entry of `package.json`. In our case, the browsers are *only* specified for development (i.e. `yarn run dev`). In the case of production, however, since there's no `production` entry, it falls back to the [reasonable set of defaults](https://browserl.ist/).

As for "Why isn't it specified?": https://github.com/c-hive/team-contribution-calendar/pull/44#discussion_r318946538.

gomorizsolt replied:
To cut a long story short, I'd say the setup is fine as-is, let's just support IE11.

gomorizsolt replied:
> let's just support IE11.

Ignore this. The setup is already fine as-is.

https://github.com/c-hive/team-contribution-calendar/blob/f6ce3a54aab963a1b4584cdc574e39fee6fa599a/.babelrc#L9-L16

We're already using `@babel/plugin-transform-runtime` to transpile generators.

thisismydesign replied:
> Target environment == "target browsers". These can be defined one-by-one by editing the browserslist entry of package.json. In our case, the browsers are only specified for development (i.e. yarn run dev). In the case of production, however, since there's no production entry, it falls back to the reasonable set of defaults.

Got it, thanks. Didn't know there's a difference between prod and dev. We should, first of all, have the same config for both where possible. Secondly, we should support the widest audience reasonably possible.

>  The setup is already fine as-is.

I'd propose to use "not dead with a usage number like > 0.2%" in both dev and prod. Is that easy to implement?

gomorizsolt replied:
> I'd propose to use "not dead with a usage number like > 0.2%" in both dev and prod. Is that easy to implement?

https://browserl.ist/?q=last+2+versions+and+not+dead+and+%3E+0.2%25

AFAICT it's as simple as using `@babel/plugin-transform-runtime` for development as well.

gomorizsolt replied:
https://github.com/c-hive/team-contribution-calendar/pull/70/commits/817acbfecf40de7af5a28009f525e179a2923349

thisismydesign commented on src/utils/CalendarUtils/CalendarUtils.js:
```diff
@@ -37,30 +38,67 @@ export const elementExists = selector => {
+export const aggregateCalendars = produce(
+  (draft, dailyDataWithContributions) => {
+    // Extract the weeks from the calendar.
+    const weeks = draft.children[0].children.slice(0, 53);
```
This comment can very easily be replaced by a function. Always think if comments are really necessary.

gomorizsolt replied:
https://github.com/c-hive/team-contribution-calendar/pull/70/commits/b7853b76f8e8b671f1f5c647ce2f0155267976a2

gomorizsolt replied:
> Always think if comments are really necessary.

Thanks for pointing this out!

thisismydesign reviewed (changes requested):
Few comments. A lot of things changed so let's make sure everything still works.

The refactor looks good :)

![image](https://user-images.githubusercontent.com/3299948/79683209-b8442d80-8228-11ea-9a05-6a1de6b037b2.png)

gomorizsolt committed:
Add extractWeeks()

Add extractWeeks() to 1) extract the weeks from the calendar and 2) remove the noise.

gomorizsolt committed:
Add Object.fromEntries()

gomorizsolt commented:
I don't know why but TCC always shows slightly fewer contributions when it's compared to GitHub:

![Screenshot 2020-04-20 at 12 51 28](https://user-images.githubusercontent.com/30864202/79744044-f82d1280-8305-11ea-8efd-03f2a17a5e8d.png)

![Screenshot 2020-04-20 at 12 51 38](https://user-images.githubusercontent.com/30864202/79744040-f6fbe580-8305-11ea-9f7f-dd760a1c1b80.png)

I've tried to debug the code but everything seems to be fine.

gomorizsolt committed:
Extract '53' to a separate variable

gomorizsolt committed:
Set minimum node version to >=12.4.0

Setting the version is required, otherwise the tests may fail because Object.fromEntries() is only supported from >=12.4.0.

gomorizsolt commented:
> A lot of things changed so let's make sure everything still works.

I've tested the changes against this config (it's supposed to cover all the scenarios):

```js
TeamContributionCalendar.default(".container", [{
    name: "gomorizsolt",
    from: "2020-01-01"
}, {
    name: "thisismydesign",
    from: "2019-09-01",
    to: "2019-09-21"
}], [{
    name: "markglenfletcher",
    to: "2019-04-30"
}])
```

Result:

![Screenshot 2020-04-21 at 8 25 01](https://user-images.githubusercontent.com/30864202/79832408-a4740500-83a9-11ea-816f-11a2ad178661.png)

thisismydesign commented on src/utils/CalendarUtils/CalendarUtils.js:
```diff
@@ -37,30 +38,66 @@ export const elementExists = selector => {
+  return Object.fromEntries(filtered);
 };
+
+export const extractWeeks = calendar => {
```
Would this make sense?

```suggestion
export const extractWeeksFromGitHub = calendar => {
```

gomorizsolt replied:
Hmmm, I think so, thanks!

gomorizsolt replied:
https://github.com/c-hive/team-contribution-calendar/pull/70/commits/e5d5689c479c020e62aff1575ffb758d0327ef80

gomorizsolt committed:
Install core-js, improve .babelrc

Merge Babel configurations to adhere to the DRY principle. Use 'core-js' instead of @babel/polyfill because the latter became deprecated as of @babel/v7.4.0.

gomorizsolt committed:
Rename extractWeeks to extractWeeksFromGitHub

gomorizsolt committed:
Extend supported browsers
